ONSdigital/eq-questionnaire-runner
# Bug fix: Backwards routing does not work when dependencies exist

### What is the context of this PR?
Fixes an issue with backwards routing due to us overwriting the reference to `routing_path_block_ids` when building the path. The variable `routing_path_block_ids` is passed to `_evaluate_routing_rules` by reference and mutated when routing backwards, however when dependent block ids exist, this reference is lost as we were re-assigning the variable: https://github.com/ONSdigital/eq-questionnaire-runner/blob/main/app/questionnaire/path_finder.py#L178-L181

This change ensures the reference is not overwritten, but instead we use a new variable and a copy of `routing_path_block_ids`. In future, we will look at rewriting pathfinder to not do such mutation which not only is unclear but prone to error as proven by this.

The updated schema simply moves the Confirmation question into the second section, which creates a when rule dependencies given the confirmation question's routing uses the first question. The additional dependencies are now populated because it is now in another section, previously it didn't need additional dependencies as it was parh of the same routing path.

### How to review
1. Use the updated test schema from this PR on `main`.
2. Get to the **Confirmation Question** (confirm-zero-employees-block) and select `No I need to correct this`
3. Confirm that on submit, the answer is removed, but you are not routed back. You should still be on the same page.
4. Checkout this branch
5. Repeat step 2
6. Confirm that on submit, you are routed back to the previous page (number-of-employees-total-block)

---

1. Test the scenario attached on the card and confirm it is no longer an issue.
2. Feel free to test any live surveys that have confirmation questions, which also has dependencies.

---
Since the issue is with a memory reference being overwritten, no new tests were needed other than updating the schema and changing some section ids in the existing test. You can ensure that the existing unit test `test_remove_answer_and_block_if_routing_backwards` is adequate but running this schema and unit test on `main` and you should see that the test fails as the routing paths do not match.

### Checklist
